File: scripts/migrate-document-vault.py
#!/usr/bin/env python
"""
Safe Migration Script for Existing Document Vault Records
Maps all existing records to service-aware namespaces without deleting any data.
"""

import sqlite3
import hashlib
import os
import re

db_path = os.path.join(os.path.dirname(__file__), '..', 'packages', 'database', 'prisma', 'dev.db')

def calculate_checksum(val):
    return hashlib.sha256(val.encode('utf-8')).hexdigest()

def sanitize_filename(filename):
    clean = os.path.basename(filename)
    clean = re.sub(r'[^a-zA-Z0-9._-]', '_', clean)
    return clean or f"doc_{int(os.times().elapsed)}"

def build_canonical_storage_key(tenant_id, service, module, entity_type, entity_id, category, filename):
    safe_tenant = re.sub(r'[^a-zA-Z0-9_-]', '_', tenant_id or 'default-tenant')
    safe_service = re.sub(r'[^a-zA-Z0-9_-]', '_', (service or 'documents').lower())
    safe_module = re.sub(r'[^a-zA-Z0-9_-]', '_', (module or 'general').lower())
    safe_entity_type = re.sub(r'[^a-zA-Z0-9_-]', '_', (entity_type or 'general').lower())
    safe_entity_id = re.sub(r'[^a-zA-Z0-9_-]', '_', entity_id or 'root')
    safe_category = re.sub(r'[^a-zA-Z0-9_-]', '_', (category or 'upload').lower())
    safe_filename = sanitize_filename(filename)
    return f"tenant/{safe_tenant}/{safe_service}/{safe_module}/{safe_entity_type}/{safe_entity_id}/{safe_category}/{safe_filename}"

def run_migration():
    print("=" * 65)
    print("🚀 MIGRATING EXISTING DOCUMENT VAULT RECORDS TO SERVICE NAMESPACES")
    print(f"Database: {db_path}")
    print("=" * 65 + "\n")

    conn = sqlite3.connect(db_path)
    c = conn.cursor()

    c.execute("SELECT id, name, originalName, mimeType, size, tenantId, checksum FROM Document")
    rows = c.fetchall()
    print(f"Audited {len(rows)} existing documents in database.")

    migrated_count = 0
    for doc in rows:
        doc_id, name, original_name, mime_type, size, tenant_id, existing_checksum = doc
        orig_name = original_name or name or "unnamed_file"

        service = "documents"
        module_name = "general"
        entity_type = "file"
        entity_id = "root"
        category = "upload"
        parent_doc_id = None

        name_lower = (name or '').lower()

        if doc_id == 'doc_resume_candidate' or 'resume' in name_lower:
            service = 'hr'
            module_name = 'employees'
            entity_type = 'candidate'
            entity_id = 'emp_syntheval_lead'
            category = 'input'
        elif name_lower.startswith('receipt_'):
            service = 'finance'
            module_name = 'invoices'
            entity_type = 'invoice'
            category = 'receipt'
            match = re.search(r'Receipt_([A-Za-z0-9_-]+)\.pdf', name, re.IGNORECASE)
            entity_id = match.group(1) if match else 'INV-TEST-2026-001'
            if entity_id == 'INV-TEST-2026-001':
                parent_doc_id = '0bb2e90b-1861-473f-8b7b-0a57f53f685c'
        elif 'invoice' in name_lower:
            service = 'finance'
            module_name = 'invoices'
            entity_type = 'invoice'
            category = 'input'
            if doc_id == 'fcff7251-81d6-4439-91b5-4eb1ac485ab9':
                entity_id = 'inv_synth_001'
            elif doc_id == '0bb2e90b-1861-473f-8b7b-0a57f53f685c':
                entity_id = 'INV-TEST-2026-001'
            else:
                entity_id = 'INV-2026-TEST'

        storage_key = build_canonical_storage_key(tenant_id, service, module_name, entity_type, entity_id, category, name)
        checksum = existing_checksum or calculate_checksum(f"{tenant_id}:{name}:{size}")

        c.execute("""
            UPDATE Document SET
                originalName = ?,
                service = ?,
                module = ?,
                entityType = ?,
                entityId = ?,
                category = ?,
                storageKey = ?,
                checksum = ?,
                parentDocumentId = ?,
                status = 'ACTIVE',
                processingStatus = 'COMPLETED',
                updatedAt = CURRENT_TIMESTAMP
            WHERE id = ?
        """, (orig_name, service, module_name, entity_type, entity_id, category, storage_key, checksum, parent_doc_id, doc_id))
        migrated_count += 1
        print(f"✓ Migrated [{doc_id}]: {service}/{module_name}/{entity_type}/{entity_id}/{category}/{name}")
        print(f"  -> Canonical Key: {storage_key}")
        print(f"  -> Checksum: {checksum[:16]}...")

    conn.commit()

    c.execute("SELECT id, name, service, module, entityId, category, storageKey, parentDocumentId FROM Document")
    verified = c.fetchall()
    print("\n" + "=" * 65)
    print(f"✅ MIGRATION COMPLETE: {len(verified)} records successfully updated and verified.")
    print("=" * 65)
    for r in verified:
        print(r)

    conn.close()

if __name__ == '__main__':
    run_migration()

File: scripts/migrate-document-vault.mjs
#!/usr/bin/env node
/**
 * Safe Migration Script for Existing Document Vault Records
 * Maps all existing records to service-aware namespaces without deleting any data.
 */

import crypto from 'crypto';
import sqlite3 from 'sqlite3';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
const dbPath = join(__dirname, '..', 'packages', 'database', 'prisma', 'dev.db');

function calculateChecksum(str) {
  return crypto.createHash('sha256').update(str).digest('hex');
}

function buildCanonicalStorageKey({ tenantId, service, module, entityType, entityId, category, filename }) {
  const safeTenant = (tenantId || 'default-tenant').replace(/[^a-zA-Z0-9_-]/g, '_');
  const safeService = (service || 'documents').toLowerCase().replace(/[^a-zA-Z0-9_-]/g, '_');
  const safeModule = (module || 'general').toLowerCase().replace(/[^a-zA-Z0-9_-]/g, '_');
  const safeEntityType = (entityType || 'general').toLowerCase().replace(/[^a-zA-Z0-9_-]/g, '_');
  const safeEntityId = (entityId || 'root').replace(/[^a-zA-Z0-9_-]/g, '_');
  const safeCategory = (category || 'upload').toLowerCase().replace(/[^a-zA-Z0-9_-]/g, '_');
  const safeFilename = filename.replace(/[^a-zA-Z0-9._-]/g, '_');

  return `tenant/${safeTenant}/${safeService}/${safeModule}/${safeEntityType}/${safeEntityId}/${safeCategory}/${safeFilename}`;
}

async function runMigration() {
  console.log('===============================================================');
  console.log('🚀 MIGRATING EXISTING DOCUMENT VAULT RECORDS TO SERVICE NAMESPACES');
  console.log(`Database: ${dbPath}`);
  console.log('===============================================================\n');

  const db = new sqlite3.Database(dbPath);

  db.all('SELECT * FROM Document', async (err, docs) => {
    if (err) {
      console.error('Failed to read Document table:', err);
      db.close();
      process.exit(1);
    }

    console.log(`Audited ${docs.length} existing documents in database.`);

    let migratedCount = 0;
    const migrations = docs.map((doc) => {
      let service = 'documents';
      let moduleName = 'general';
      let entityType = 'file';
      let entityId = 'root';
      let category = 'upload';
      let parentDocumentId = null;

      const name = doc.name || '';
      const originalName = doc.originalName || name;

      // Classify existing documents
      if (doc.id === 'doc_resume_candidate' || name.toLowerCase().includes('resume')) {
        service = 'hr';
        moduleName = 'employees';
        entityType = 'candidate';
        entityId = 'emp_syntheval_lead';
        category = 'input';
      } else if (name.toLowerCase().startsWith('receipt_')) {
        service = 'finance';
        moduleName = 'invoices';
        entityType = 'invoice';
        category = 'receipt';
        
        // Extract invoice number from name
        const match = name.match(/Receipt_([A-Za-z0-9_-]+)\.pdf/i);
        entityId = match ? match[1] : 'INV-TEST-2026-001';

        // Link to parent invoice if matching
        if (entityId === 'INV-TEST-2026-001') {
          parentDocumentId = '0bb2e90b-1861-473f-8b7b-0a57f53f685c';
        }
      } else if (name.toLowerCase().includes('invoice')) {
        service = 'finance';
        moduleName = 'invoices';
        entityType = 'invoice';
        category = 'input';

        if (doc.id === 'fcff7251-81d6-4439-91b5-4eb1ac485ab9') {
          entityId = 'inv_synth_001';
        } else if (doc.id === '0bb2e90b-1861-473f-8b7b-0a57f53f685c') {
          entityId = 'INV-TEST-2026-001';
        } else {
          entityId = 'INV-2026-TEST';
        }
      }

      const storageKey = buildCanonicalStorageKey({
        tenantId: doc.tenantId,
        service,
        module: moduleName,
        entityType,
        entityId,
        category,
        filename: name,
      });

      const checksum = doc.checksum || calculateChecksum(`${doc.tenantId}:${doc.name}:${doc.size}`);

      return {
        id: doc.id,
        name,
        originalName,
        service,
        module: moduleName,
        entityType,
        entityId,
        category,
        storageKey,
        checksum,
        parentDocumentId,
        tenantId: doc.tenantId,
      };
    });

    db.serialize(() => {
      const stmt = db.prepare(`
        UPDATE Document SET
          originalName = ?,
          service = ?,
          module = ?,
          entityType = ?,
          entityId = ?,
          category = ?,
          storageKey = ?,
          checksum = ?,
          parentDocumentId = ?,
          status = 'ACTIVE',
          processingStatus = 'COMPLETED',
          updatedAt = CURRENT_TIMESTAMP
        WHERE id = ?
      `);

      for (const m of migrations) {
        stmt.run(
          m.originalName,
          m.service,
          m.module,
          m.entityType,
          m.entityId,
          m.category,
          m.storageKey,
          m.checksum,
          m.parentDocumentId,
          m.id,
          function (updateErr) {
            if (updateErr) {
              console.error(`Error migrating doc ${m.id}:`, updateErr);
            } else {
              migratedCount++;
              console.log(`✓ Migrated [${m.id}]: ${m.service}/${m.module}/${m.entityType}/${m.entityId}/${m.category}/${m.name}`);
              console.log(`  -> Canonical Storage Key: ${m.storageKey}`);
              console.log(`  -> Checksum: ${m.checksum.substring(0, 16)}...`);
            }
          }
        );
      }

      stmt.finalize(() => {
        db.all('SELECT id, name, service, module, entityId, category, storageKey, parentDocumentId FROM Document', (checkErr, updatedRows) => {
          console.log('\n===============================================================');
          console.log(`✅ MIGRATION COMPLETE: ${updatedRows.length} records verified.`);
          console.table(updatedRows);
          db.close();
        });
      });
    });
  });
}

runMigration();
